retro/rw_nocache_md.asm

```diff
@@ -0,0 +1,427 @@
+	jr	z,.bios_write_sdbuf
+
+	call	iputs
+	db	"BIOS_WRITE SD CARD PRE-READ FAILED!\r\n\0"
+	ld	a,1			; tell CP/M the read failed
+	ld	(.bios_sdbuf_val),a	; mark the .bios_sdbuf as invalid
+	jp	.bios_write_ret
+
+.bios_write_sdbuf:
+	; calculate the CP/M sector offset address (bios_disk_sector*128)
+	ld	hl,(bios_disk_sector)	; must be 0..3
+	add	hl,hl			; HL *= 2
+	add	hl,hl			; HL *= 4
+	add	hl,hl			; HL *= 8
+	add	hl,hl			; HL *= 16
+	add	hl,hl			; HL *= 32
+	add	hl,hl			; HL *= 64
+	add	hl,hl			; HL *= 128
+
+	; calculate the address of the CP/M sector in the .bios_sdbuf
+	ld	bc,.bios_sdbuf
+	add	hl,bc			; HL = @ of cpm sector in the .bios_sdbuf
+	ld	d,h
+	ld	e,l			; DE = @ of cpm sector in the .bios_sdbuf
+
+	; copy the data of interest /into/ the SD block
+	ld	hl,(bios_disk_dma)		; source address
+	ld	bc,0x0080		; number of bytes to copy
+	ldir
+
+	; write the .bios_sdbuf contents to the SD card
+	ld      hl,(bios_disk_track)
+	ld	de,.sd_partition_base	; XXX add the starting partition block number
+	add	hl,de			; HL = SD physical block number
+	ld	de,0
+	call	.bios_disk_offset	;calculate disk offset based on disk number
+	push	de			; SD block number to write
+	push	hl
+	ld	de,.bios_sdbuf		; DE = target buffer to read the 512-byte block
+	call	sd_cmd24		; write the SD block
+	pop	hl			; clean the SD block number from the stack
+	pop	de
+
+	or	a
+	jr	z,.bios_write_ret
+
+	call	iputs
+	db	"BIOS_WRITE SD CARD WRITE FAILED!\r\n\0"
+	ld	a,1			; tell CP/M the read failed
+	ld	(.bios_sdbuf_val),a	; mark the .bios_sdbuf as invalid
+
+.bios_write_ret:
+	pop	bc
+	pop	de			; restore saved regs
+
+	pop	hl			; HL = original saved stack pointer
+	ld	sp,hl			; SP = original stack address
+	pop	hl			; restore the original  HL value
+
+	ret
+
+
+;##########################################################################
+; A single SD block cache
+;##########################################################################
+.bios_sdbuf_trk:		; The CP/M track number last left in the .bios_sdbuf
+	ds	2,0xff		; initial value = garbage
+.bios_sdbuf_val:		; The CP/M track number in .bios_sdbuf_trk is valid when this is 0
+	ds	1,0xff		; initial value = INVALID
+.bios_sdbuf:			; scratch area to use for SD block reading and writing
+	ds	512,0xa5	; initial value = garbage
+
+
+;##########################################################################
+; Called once before library is used.
+;##########################################################################
+rw_init:
+;	call	iputs
+;	db	'NOTICE: rw_nocache library installed. Disk cache disabled.\r\n\0'
+
+	ld	a,1
+	ld	(.bios_sdbuf_val),a     ; mark .bios_sdbuf_trk as invalid
+
+        ret
+        
+;##########################################################################
+;Modification for multiple disks - Trevor Jacobs - 02-15-2023
+;add 0x4000 to de,hl disk number of times - 0x4000 offset per disk
+;##########################################################################
+.bios_disk_offset:
```

Comment:
I recommend changing how this function operates:
input:
 _(bios_disk_track)_ ;cpm track number 0x0..0x3fff
 _(bios_disk_current_disk)_ ;0x0..0x3
output:
 _DE:HL_ physical SdCard sector

every time you call this function you load _HL_ from _(bios_disk_track)_, add _.sd_partition_base_, and Zero out _DE_.  Do all this work in this function.

It would really be nice if _.sd_partion_base_ was a 32bit sector number read from the Partition table of the SdCard, instead hard coded at 0x800.